### ISSUE TITLE
chezmoi edit with neovim hangs after making edits

### ISSUE THREAD

**Issue by jgoguen** [twpayne/chezmoi#3885]
## Describe the bug

When using chezmoi 2.51.0 and neovim 0.10.1 to edit files, exiting neovim after making edits hangs for some time (many seconds at least) then exits and usually leaves the terminal in a poor state (`reset` is needed to make it usable again).

I've reproduced this with the macOS ARM64 binary from both the Releases page and installed from Homebrew. I'm using ZSH form Homebrew and I've reproduced this in the current version of both Kitty and WezTerm and the macOS terminal. I'm using macOS 14.6.

I have also reproduced this on Fedora 40, default GNOME Shell, with GNOME Terminal and Kitty (also WezTerm, but I'm having other issues with WezTerm on Fedora so although it seems likely to be chezmoi considering it's happening across multiple OSes and terminals, I can't say for sure it's only chezmoi there), also using ZSH.

Looking at the process tree, when first running `chezmoi edit` it seems to be running neovim as expected (sample here is from `pstree` on macOS, but Fedora is similar):

```console
 | | \-+= 19185 jgoguen chezmoi edit --verbose -a /Users/jgoguen/.zaliases
 | |   |-+- 19186 jgoguen /opt/homebrew/bin/nvim /Users/jgoguen/.local/share/chezmoi/dot_zaliases.tmpl
 | |   | \--= 19187 jgoguen /opt/homebrew/bin/nvim --embed /Users/jgoguen/.local/share/chezmoi/dot_zaliases.tmpl
```

When exiting, it seems chezmoi is trying to run my pager for some reason.

```console
 | | \-+= 19185 jgoguen chezmoi edit --verbose -a /Users/jgoguen/.zaliases
 | |   |-+- 19186 jgoguen /opt/homebrew/bin/nvim /Users/jgoguen/.local/share/chezmoi/dot_zaliases.tmpl
 | |   | \--= 19187 jgoguen /opt/homebrew/bin/nvim --embed /Users/jgoguen/.local/share/chezmoi/dot_zaliases.tmpl
 | |   \--- 20327 jgoguen less -FRXd
```

Killing the `less` process does not resolve anything, the child is left as a zombie process.

If I `unset PAGER`, chezmoi still hangs after exiting neovim but no additional process is started. The terminal is usually not left in an unusable state though.

## To reproduce

- Use chezmoi to edit a managed file (e.g. `chezmoi edit -a ~/.zshrc` or `chezmoi edit ~/.bash_profile`). Using `-a` or not doesn't change the behaviour.
- Make a change to the file (this does not reproduce if no actual changes are made) and exit neovim

## Expected behavior

Chezmoi exits cleanly, pausing only to apply the changed file if `-a` was given.

## Output of command with the `--verbose` flag

```console
$ chezmoi edit --verbose -a ~/.zshrc
(no output)
```

## Output of `chezmoi doctor`

```console
# macOS
% chezmoi doctor
RESULT   CHECK                       MESSAGE
ok       version                     v2.51.0, commit Homebrew, built at 2024-07-15T19:36:30Z, built by Homebrew
ok       latest-version              v2.51.0
ok       os-arch                     darwin/arm64
ok       uname                       Darwin atlas 23.6.0 Darwin Kernel Version 23.6.0: Fri Jul  5 17:56:41 PDT 2024; root:xnu-10063.141.1~2/RELEASE_ARM64_T6000 arm64
ok       go-version                  go1.22.5 (gc)
ok       executable                  /opt/homebrew/bin/chezmoi
ok       upgrade-method              brew-upgrade
ok       config-file                 ~/.config/chezmoi/chezmoi.toml, last modified 2024-03-20T09:21:10-04:00
ok       source-dir                  ~/.local/share/chezmoi is a git working tree (clean)
ok       suspicious-entries          no suspicious entries
ok       working-tree                ~/.local/share/chezmoi is a git working tree (clean)
ok       dest-dir                    ~ is a directory
ok       umask                       022
ok       cd-command                  found /opt/homebrew/bin/zsh
ok       cd-args                     /opt/homebrew/bin/zsh -l
info     diff-command                not set
ok       edit-command                found /opt/homebrew/bin/nvim
ok       edit-args                   /opt/homebrew/bin/nvim
ok       git-command                 found /opt/homebrew/bin/git, version 2.46.0
ok       merge-command               found /usr/bin/vimdiff
ok       shell-command               found /opt/homebrew/bin/zsh
ok       shell-args                  /opt/homebrew/bin/zsh
ok       age-command                 found /opt/homebrew/bin/age, version 1.2.0
ok       gpg-command                 found /opt/homebrew/bin/gpg, version 2.4.5
info     pinentry-command            not set
ok       1password-command           found /opt/homebrew/bin/op, version 2.30.0
info     bitwarden-command           bw not found in $PATH
info     bitwarden-secrets-command   bws not found in $PATH
info     dashlane-command            dcli not found in $PATH
info     doppler-command             doppler not found in $PATH
info     gopass-command              gopass not found in $PATH
info     keepassxc-command           keepassxc-cli not found in $PATH
info     keepassxc-db                not set
info     keeper-command              keeper not found in $PATH
info     lastpass-command            lpass not found in $PATH
info     pass-command                pass not found in $PATH
info     passhole-command            ph not found in $PATH
info     rbw-command                 rbw not found in $PATH
info     vault-command               vault not found in $PATH
info     vlt-command                 vlt not found in $PATH
info     secret-command              not set

# Fedora
% ./chezmoi-linux-amd64 doctor
RESULT   CHECK                       MESSAGE
ok       version                     v2.51.0, commit 2a7845f4ffe2fd427c140711c0d7d46424747363, built at 2024-07-15T20:05:05Z, built by goreleaser
ok       latest-version              v2.51.0
ok       os-arch                     linux/amd64 (Fedora Linux 40 (Server Edition))
ok       uname                       Linux mori 6.9.7-200.fc40.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Jun 27 18:11:45 UTC 2024 x86_64 GNU/Linux
ok       go-version                  go1.22.5 (gc)
ok       executable                  ~/chezmoi-linux-amd64
ok       upgrade-method              replace-executable
ok       config-file                 ~/.config/chezmoi/chezmoi.toml, last modified 2024-07-29T22:37:39-04:00
ok       source-dir                  ~/.local/share/chezmoi is a git working tree (clean)
ok       suspicious-entries          no suspicious entries
ok       working-tree                ~/.local/share/chezmoi is a git working tree (clean)
ok       dest-dir                    ~ is a directory
ok       umask                       022
ok       cd-command                  found /usr/bin/zsh
ok       cd-args                     /usr/bin/zsh -l
info     diff-command                not set
ok       edit-command                found /usr/bin/nvim
ok       edit-args                   /usr/bin/nvim
ok       git-command                 found /usr/bin/git, version 2.45.2
ok       merge-command               found /usr/bin/vimdiff
ok       shell-command               found /usr/bin/zsh
ok       shell-args                  /usr/bin/zsh
info     age-command                 age not found in $PATH
ok       gpg-command                 found /usr/bin/gpg, version 2.4.4
info     pinentry-command            not set
ok       1password-command           found /usr/bin/op, version 2.30.0
info     bitwarden-command           bw not found in $PATH
info     bitwarden-secrets-command   bws not found in $PATH
info     dashlane-command            dcli not found in $PATH
info     doppler-command             doppler not found in $PATH
info     gopass-command              gopass not found in $PATH
info     keepassxc-command           keepassxc-cli not found in $PATH
info     keepassxc-db                not set
info     keeper-command              keeper not found in $PATH
info     lastpass-command            lpass not found in $PATH
info     pass-command                pass not found in $PATH
info     passhole-command            ph not found in $PATH
info     rbw-command                 rbw not found in $PATH
info     vault-command               vault not found in $PATH
info     vlt-command                 vlt not found in $PATH
info     secret-command              not set
```

</details>

## Additional context

This also seems to be happening with chezmoi 2.50. Sometimes I catch a flash of text just as neovim exits, sometimes there's nothing, but no messages or errors appear anywhere I can read them. All the changes do seem to be saved, and applied if I pass `-a`, the only problem appears to be actually exiting neovim. This does not reproduce with neovim alone.

**Comment by halostatue** [comment]
If your terminal allows you to see the "alternate screen" from a menu, check that to see if something was printed there (it should still be there). On iTerm2, this is reachable with `Session | Terminal State | Alternate Screen`. There are `tput` commands available and some escape codes that work on xterms (see this [SO answer](https://stackoverflow.com/questions/11023929/using-the-alternate-screen-in-a-bash-script#11024208)).

I do not use neovim by default and use Fish by default, so I cannot reproduce your environment exactly, but across four terminals (wezterm, kitty, alacritty, and iterm2) and both neovim 0.10.0 and 0.10.1 I did the following:

```console
fish$ zsh
zsh$ VISUAL=nvim chezmoi edit --verbose --apply ~/.config/git/config
# make edits
# ZZ
```

I could not produce the hanging behaviour you report nor did I see any flashes of text afterwards‡. Nor could I reproduce it with vim 9.1.

Do you use [chezmoi.vim](https://github.com/alker0/chezmoi.vim)? If so, consider disabling it and see if the behaviour continues (I don't think this would be the source of the problem, but it's worth trying). This would also apply to [GianniBYoung/chezmoi-telescope.nvim](https://github.com/GianniBYoung/chezmoi-telescope.nvim).

I think that there is something else going on with your system and I’d be more inclined to see it within your neovim config than in chezmoi, as nothing in the last several releases seems to have modified code surrounding `chezmoi edit`.

‡ I had warnings pop up in my poorly managed neovim configuration because (as I said) I don't use neovim after many attempts and failures to find it an acceptable vim alternative. These warnings are not related to chezmoi.

**Comment by jgoguen** [comment]
Apologies, I thought I had set neovim to use a different config dir while trying to track this down but apparently I did not. This time I actually moved my config out of the way and I don't see this anymore. I don't have `chezmoi.vim`, nor does the alternate screen show anything, but since I know this is something to do with my setup I can pare it down and start adding things back until it breaks again.
